File: ID10189_Minesweeper.py
# @JUDGE_ID: 10189 Python "Minesweeper" 

# @BEGIN_OF_SOURCE_CODE
from sys import stdin

def checkField(FIELD, row, col):
	nrow = [row-1,row-1,row-1,row,row,row+1,row+1,row+1]
	ncol = [col-1,col,col+1,col-1,col+1,col-1,col,col+1]

	M = len(FIELD)
	N = len(FIELD[0])-1

	nmina = 0
	for i in range(len(nrow)):
		if nrow[i] >= 0 and nrow[i] <M:
			if ncol[i] >= 0 and ncol[i] <N:
				if FIELD[nrow[i]][ncol[i]] =='*':
					nmina+=1
	return nmina

def Minesweeper(FIELD,M,N):
	for i in range(0,M):
		for j in range(0,N):
			if FIELD[i][j] == '.':
				nmina = checkField(FIELD,i,j)
				FIELD[i][j]=  str(nmina)



if __name__ == '__main__':
	nfield = 1
	status = False
	while(True):
		FIELD = []
		line = stdin.readline()
		v = line.split()
		M = int(v[0])
		N = int(v[1])
		if M==0 and N==0:
			break
		for x in range(M):
			line = stdin.readline()
			FIELD.append(list(line))
		
		Minesweeper(FIELD,M,N)
		if status:
			print()
		else:
			status = True
		print("Field #"+str(nfield)+":")
		for x in range(0,M):
			print(''.join(FIELD[x]),end='')

		nfield+=1






#@END_OF_SOURCE_CODE*/

File: ID10189_Minesweeper.cpp
/* @JUDGE_ID: 10189 C++ "Minesweeper" */

/* @BEGIN_OF_SOURCE_CODE*/
#include <stdio.h>
#include<iostream>
#include <stdlib.h> /*Character handling functions*/
#include<string.h>


using namespace std;

const int INF = 0x3F3F3F3F;
const int NULO = -1;
const double EPS = 1e-10;

#define MAX 102
char FIELD[MAX][MAX];
char FIELDN[MAX][MAX];

int M,N;

int checkField(int row, int col){
	int i;
	int nmina = 0;
	int index[8][2];

	index[0][0] = row-1;
	index[1][0] = row-1;
	index[2][0] = row-1;
	index[3][0] = row;
	index[4][0] = row;
	index[5][0] = row+1;
	index[6][0] = row+1;
	index[7][0] = row+1;

	index[0][1] = col-1;
	index[1][1] = col;
	index[2][1] = col+1;
	index[3][1] = col-1;
	index[4][1] = col+1;
	index[5][1] = col-1;
	index[6][1] = col;
	index[7][1] = col+1;

	for(i=0; i<8;i++){
		if (index[i][0]>=0 && index[i][0]< M )
			if (index[i][1]>=0 && index[i][1]< N )
				if (FIELD[index[i][0]][index[i][1]]=='*')
					nmina++;
	}

	return nmina;
}

int Minesweeper(){
	int i,j;

	for(i=0;i<M;i++){
		FIELDN[i][N] = '\0';
		for(j=0;j<N;j++){
			if(FIELD[i][j] == '*'){//tem mina
				FIELDN[i][j] = '*';

			}else{//
				FIELDN[i][j] =  checkField(i,j)+ '0';

			}
		}

	}

	return 0;
}


int main() {
	int i,nn,status;
	string line;

	nn=1;
	status = 0;
	while (true){
		cin>>M>>N;
		if (M==0 && N==0)
			break;

		for(i=0;i<M;i++)
			scanf("%s\n",FIELD[i]);
			//cin>>line;
			//strcpy(FIELD[i],line.c_str());
		
		Minesweeper();
		if (status)
				printf("\n");
		else
			status = 1;
		printf("Field #%d:\n",nn);
		for(i=0;i<M;i++)
			printf("%s\n",FIELDN[i]);

		nn++;
		
	}
 
  return 0;
}

/*@END_OF_SOURCE_CODE*/